src/Expander.hs

```diff
@@ -513,14 +513,14 @@ initializeKernel = do
                 then throwError $ WrongDatatypeArity stx dt arity (length args)
                 else do
                   varInfo <- traverse Prims.prepareVar args
-                  sch <- trivialScheme $ Ty TType
+                  sch <- trivialScheme tType
                   modifyState $
                     set (expanderPatternBinders . at (Right dest)) $
                     Just [ (sc, n, x, sch)
                          | (sc, n, x) <- varInfo
                          ]
                   linkTypePattern dest $
-                    TypePattern $ TDatatype dt [(varStx, var) | (_, varStx, var) <- varInfo]
+                    TypePattern $ TyF (TDatatype dt) [(varStx, var) | (_, varStx, var) <- varInfo]
```

Comment:
The diffs in this file up to this point illustrate a few different situations in which `TyF` is used. Some functions ask for a `Ty`, which wraps a `TyF Ty`, while other functions as for a `TyF a` for an `a` which isn't `Ty`. Those were getting pretty long an unreadable, so I decided to define a bunch of helper functions, such as `tFun` and `tSyntax`, for constructing such values more easily.

I had a design choice: should I make those helpers for the `Ty` or the `TyF` variant? Making them for the `TyF` variant would improve all the call sites, at the cost of extra `Ty` wrappers in the places where `Ty` is needed. I decided instead fo make them for the `Ty` variant, because `Ty` expressions like the one on L396 are long and nested, and so the extra readability helps a lot more than in the `TyF` expressions, which tend to only be one level deep.

Reply:
Sounds good.

What about making the helpers into a type class that can have instances for both cases?

Reply:
good idea! pushed.